### PR TITLE
Use Microsoft.DiaSymReader.Native to make PDB id and stamp deterministic

### PR DESCRIPTION
**Pull request by tmat**

### CODE REVIEW

**Comment by tmat** [comment]
@gafter @jaredpar @pharring @jasonmalinowski

File: src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs [hunk 1]
```diff
@@ -620,8 +682,31 @@ public void SetMetadataEmitter(MetadataWriter metadataWriter)
...
 
-        public unsafe void GetDebugDirectoryGuidAndStampAndAge(out Guid guid, out uint stamp, out uint age)
+        public unsafe ContentId GetContentId()
         {
+            if (_deterministic)
+            {
+                // Call to GetDebugInfo fails for SymWriter initialized using InitializeDeterministic.
+                // We already have all the info we need though.
+
+                // TODO: calculate sha1 hash
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs]: bug number?

File: src/Compilers/Core/Portable/PEWriter/ContentId.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+
+namespace Microsoft.Cci
+{
+    internal struct ContentId
+    {
+        public readonly byte[] Guid;
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/PEWriter/ContentId.cs]: Is `ImmutableArray<byte>` appropriate here?

File: src/Compilers/Core/Portable/PEWriter/PeWriter.cs [hunk 3]
```diff
@@ -1487,10 +1437,10 @@ private void WriteDebugTable(Stream peStream, MetadataSizes metadataSizes, out l
             writer.WriteByte((byte)'S');
 
             // PDB id:
-            writer.WriteBytes(pdbId.ToByteArray());
+            writer.WriteBytes(pdbContentId.Guid);
 
             // age
-            writer.WriteUint(age);
+            writer.WriteUint(1);
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/PEWriter/PeWriter.cs]: 1 pops up as a literal in a few places.  Can we make it a named constant?

File: src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs [hunk 4]
```diff
@@ -586,22 +590,67 @@ private void DefineScopeLocals(LocalScope currentScope, uint localSignatureToken
...
             {
                 // If an exception is thrown we propagate it - we want to report it every time. 
-                s_lazyCorSymWriterSxSType = Marshal.GetTypeFromCLSID(new Guid("0AE2DEB0-F901-478b-BB9F-881EE8066788"));
+                s_lazyCorSymWriterSxSType = Marshal.GetTypeFromCLSID(new Guid(SymWriterClsid));
             }
 
             return s_lazyCorSymWriterSxSType;
         }
 
+        private object CreateSymWriterWorker()
```
Comment (pharring) [inline src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs]: static

**Comment by pharring** [comment]
LGTM

File: src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs]: https://github.com/dotnet/roslyn/issues/926

Comment (gafter) [inline src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs]: See also https://github.com/gafter/roslyn/pull/3 which is a start on the code that will fill in the hole here.

File: src/Compilers/Core/Portable/PEWriter/ContentId.cs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+        internal static ContentId FromSha1(ImmutableArray<byte> sha1)
+        {
+            var guid = new byte[16];
+            for (var i = 0; i < guid.Length; i++)
+            {
+                guid[i] = sha1[i];
+            }
+
+            // modify the guid data so it decodes to the form of a "random" guid ala rfc4122
```
Comment (gafter) [inline src/Compilers/Core/Portable/PEWriter/ContentId.cs]: This looks suspiciously copy-and-paste. Please see my https://github.com/gafter/roslyn/pull/3 which moves it to a common API.

File: src/Compilers/Core/Portable/PEWriter/ContentId.cs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+        public ContentId(byte[] guid, byte[] stamp)
+        {
+            Debug.Assert(guid.Length == 16 && stamp.Length == 4);
+
+            Guid = guid;
+            Stamp = stamp;
+        }
+
+        internal static ContentId FromSha1(ImmutableArray<byte> sha1)
```
Comment (gafter) [inline src/Compilers/Core/Portable/PEWriter/ContentId.cs]: FromHash, as it should work with any hash that provides enough bytes.

File: src/Compilers/Core/Portable/PEWriter/PeWriter.cs [hunk 7]
```diff
@@ -115,6 +88,8 @@ private bool WritePeToStream(MetadataWriter mdWriter, Func<Stream> getPeStream,
             var managedResourceBuffer = new MemoryStream(1024);
             var managedResourceWriter = new BinaryWriter(managedResourceBuffer);
 
+            nativePdbWriterOpt?.SetMetadataEmitter(mdWriter);
```
Comment (gafter) [inline src/Compilers/Core/Portable/PEWriter/PeWriter.cs]: What is going on here? Why does the native PDB writer need mdWriter?

File: src/Compilers/Core/Portable/PEWriter/PeWriter.cs [hunk 8]
```diff
@@ -198,37 +199,34 @@ private int CalculateMappedFieldDataStreamRva(MetadataSizes metadataSizes)
...
-            stream.Position = positionOfHeaderTimestamp;
-            stream.Write(timestamp, 0, 4);
-            if (positionOfDebugTableTimestamp != 0)
-            {
-                CheckZeroDataInStream(stream, positionOfDebugTableTimestamp, 4);
-                stream.Position = positionOfDebugTableTimestamp;
-                stream.Write(timestamp, 0, 4);
-            }
+            peStream.Position = 0;
+            var contentId = ContentId.FromSha1(CryptographicHashProvider.ComputeSha1(peStream));
```
Comment (gafter) [inline src/Compilers/Core/Portable/PEWriter/PeWriter.cs]: Oh, you consolidated the code too, just in a different place. Good.

File: src/Compilers/Core/Portable/PEWriter/ContentId.cs (hunk 2 above)
Comment (gafter) [inline src/Compilers/Core/Portable/PEWriter/ContentId.cs]: It is much easier to (1) copy the data into a stream, and (2) make a Guid, from this format.

**Comment by gafter** [comment]
LGTM

File: src/Compilers/Core/Portable/PEWriter/PeWriter.cs (hunk 7 above)
Comment (tmat) [inline src/Compilers/Core/Portable/PEWriter/PeWriter.cs]: Unfortunately it uses it to get names for methods and types :( It's dumb, but it is what it is.

File: src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs (hunk 1 above)
Comment (tmat) [inline src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs]: Fixed.

File: src/Compilers/Core/Portable/PEWriter/PeWriter.cs (hunk 3 above)
Comment (tmat) [inline src/Compilers/Core/Portable/PEWriter/PeWriter.cs]: Fixed.

File: src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs (hunk 4 above)
Comment (tmat) [inline src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs]: Fixed.

**Comment by tmat** [comment]
@dotnet-bot retest this, please.

File: src/Compilers/Core/Portable/PEWriter/PeWriter.cs (hunk 7 above)
Comment (gafter) [inline src/Compilers/Core/Portable/PEWriter/PeWriter.cs]: OK, then my logger (#926) will have to account for that input to the PDB writer. Ugh. How do I log those callbacks?

Comment (tmat) [inline src/Compilers/Core/Portable/PEWriter/PeWriter.cs]: The callbacks are asking for the name of a type given its token, for the name of a method given a token and for the containing type token given a nested type token.

If we log the name of the method and the qualified name of its containing type in OpenMethod we should be good.

**Pull request merged**